### PR TITLE
runtime: Enable all the "non dangerous" annotations by default

### PR DESCRIPTION
**Pull request by fidencio**
Let's have (almost) all the annotations enabled by default, with the
exceptions being the binary paths that are executed on the host side, in
order to avoid falling into the CVE-2020-27151, which was the reason we
disabled all the annotations in the first place.

Reference:
https://nvd.nist.gov/vuln/detail/CVE-2020-27151

Fixes: #2090

Signed-off-by: Fabiano Fidêncio <fidencio@redhat.com>

### CODE REVIEW

**Comment by eadamsintel** [comment]
It would be helpful if the order was listed in the same order as the[ Annotation page](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-set-sandbox-config-kata.md). It isn't critical. It would also help to list which ones are not enabled but not completely necessary.

Below is the list of ones I found that were not being enabled when comparing the lists. 

**io.katacontainers.config_path
io.katacontainers.pkg.oci.bundle_path
io.katacontainers.pkg.oci.container_type
io.katacontainers.config.hypervisor.disable_image_nvdimm**

This one seems like it might be ok to enable.  I can't think of how a pod stating they want to use an nvdimm as a backend could compromise the host. It might cause the pod to not start

**io.katacontainers.config.hypervisor.enable_virtio_mem**

If a pod can enable virtio-mem for their launch of Kata does that increase risk to the host at all?  I don't know enough about virtio-mem to understand why. It seems like all it is doing is passing additional QEMU parameters vs modifying paths on the host. 

**io.katacontainers.config.hypervisor.hypervisor_hash**

I don't know what this behavior is but if I am understanding the description then if the hash fails my guess is the Pod wouldn't start. If that is how it works then wouldn't enabling it allow pods the ability to add extra verification security before running a different hypervisor?

**io.katacontainers.config.hypervisor.jailer_hash**

Similar comment as above. Is the hash adding an extra step of integrity measurement before starting a pod? I don't understand the behavior well enough to know why or how it could compromise a host. 

**io.katacontainers.config.hypervisor.path
io.katacontainers.config.hypervisor.use_vsock**

Does allowing a pod to use Vsock on their pod instance open up the host to attacks? 

**io.katacontainers.config.hypervisor.virtio_fs_daemon**


All of the options that looked ok above to me seem pretty advanced so not having them as a default probably wouldn't be too inconvenient to anyone. If you want to go with the list as it is I am ok with that.

**Review by c3d — APPROVED** [review]
I like the idea of enabling more annotations.
Could you add a comment list the ones you excluded, and why?

**Comment by fidencio** [comment]
> It would be helpful if the order was listed in the same order as the[ Annotation page](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/how-to-set-sandbox-config-kata.md). It isn't critical. It would also help to list which ones are not enabled but not completely necessary.

Ack, I'll modify the PR accordingly. 

> Below is the list of ones I found that were not being enabled when comparing the lists.
> 
> **io.katacontainers.config_path io.katacontainers.pkg.oci.bundle_path io.katacontainers.pkg.oci.container_type io.katacontainers.config.hypervisor.disable_image_nvdimm**
> 
> This one seems like it might be ok to enable. I can't think of how a pod stating they want to use an nvdimm as a backend could compromise the host. It might cause the pod to not start

Those seen like a mislook from my side.

> 
> **io.katacontainers.config.hypervisor.enable_virtio_mem**
> 
> If a pod can enable virtio-mem for their launch of Kata does that increase risk to the host at all? I don't know enough about virtio-mem to understand why. It seems like all it is doing is passing additional QEMU parameters vs modifying paths on the host.

Same here.

> 
> **io.katacontainers.config.hypervisor.hypervisor_hash**
> 
> I don't know what this behavior is but if I am understanding the description then if the hash fails my guess is the Pod wouldn't start. If that is how it works then wouldn't enabling it allow pods the ability to add extra verification security before running a different hypervisor?
> 
> **io.katacontainers.config.hypervisor.jailer_hash**
> 
> Similar comment as above. Is the hash adding an extra step of integrity measurement before starting a pod? I don't understand the behavior well enough to know why or how it could compromise a host.

Those two were not enabled because I'm not enabling the jailer and hypervisor path annotations.
I thought those would go together.

Do you see it differently?  Do you see a reason for someone setting the hash without modifying the jailer / hypervisor path?

> 
> **io.katacontainers.config.hypervisor.path io.katacontainers.config.hypervisor.use_vsock**
> 
> Does allowing a pod to use Vsock on their pod instance open up the host to attacks?

Ah, vsock. :-).  It should be enabled and seems like another one that I missed.
However, hypervisor_path shouldn't.

> 
> **io.katacontainers.config.hypervisor.virtio_fs_daemon**

Same about this one, it shouldn't be enabled.

> 
> All of the options that looked ok above to me seem pretty advanced so not having them as a default probably wouldn't be too inconvenient to anyone. If you want to go with the list as it is I am ok with that.

Thanks for the review, I'll ping you for a look at the v2 at some point soon!

**Comment by fidencio** [comment]
> I like the idea of enabling more annotations.
> Could you add a comment list the ones you excluded, and why?

I will do as part of the next iteration, thanks for the suggestion!

**Comment by eadamsintel** [comment]
> > **io.katacontainers.config.hypervisor.hypervisor_hash**
> > I don't know what this behavior is but if I am understanding the description then if the hash fails my guess is the Pod wouldn't start. If that is how it works then wouldn't enabling it allow pods the ability to add extra verification security before running a different hypervisor?
> > **io.katacontainers.config.hypervisor.jailer_hash**
> > Similar comment as above. Is the hash adding an extra step of integrity measurement before starting a pod? I don't understand the behavior well enough to know why or how it could compromise a host.
> 
> Those two were not enabled because I'm not enabling the jailer and hypervisor path annotations.
> I thought those would go together.
> 
> Do you see it differently? Do you see a reason for someone setting the hash without modifying the jailer / hypervisor path?

I don't really know what these are tbh.  Does the hash and the path work together? Is there any scenario where you would use the hash but not the path?  Maybe you want extra validation that the host didn't upgrade / change the installed hypervisor.  Something like this is pretty advanced it seems so it might be ok to wait for an actual user to complain it should be enabled if you decide not to automatically.

My formatting might have been confusing, because I was listing all the ones I saw you didn't enable and only commenting on the ones that I thought could be enabled.

**Review by devimc — APPROVED** [review]
thanks @fidencio

File: src/runtime/Makefile [hunk 1]
```diff
@@ -159,7 +159,66 @@ DEFMEMSZ := 2048
 DEFMEMSLOTS := 10
 #Default number of bridges
 DEFBRIDGES := 1
-DEFENABLEANNOTATIONS := []
+DEFENABLEANNOTATIONS := [			\
```
Comment (jodh-intel) [inline src/runtime/Makefile:162]: :skull_and_crossbones: :radioactive: ? This has to be the scariest / coolest PR title we've had in a while! :smile:

How about tweaking slightly to something like:

```make
# Comments explaining:
#   1. What constitutes "non-dangerous".
#   2. When this should be updated (do we need a comment in `annotations.go` btw?)
#   3. The impact of values not specified in this list.
NON_DANGEROUS_ANNOTATIONS := [ ... ]

DEFENABLEANNOTATIONS := $(NON_DANGEROUS_ANNOTATIONS)
```

Also, could you sort that list alphabetically?

I'm vaguely wondering whether we should auto-generate `virtcontainers/pkg/annotations/annotations.go` and this list from a YAML file maybe.

**Review by jodh-intel — COMMENTED** [review]
Thanks @fidencio. A few comments.

**Comment by amshinde** [comment]
@fidencio What if we have a list of disabled annotations and filter against those rather than have a list of allowed annotations?

**Comment by jodh-intel** [comment]
I think it's "safer" to maintain a list of allowed annotations since any new ones that we forgot to add to this list would be automatically denied (which would be safer than automatically allowing). That said, maybe this is a good topic for the AC meeting as it is a sensitive one?

**Comment by fidencio** [comment]
@amshinde, I thought about and even opened https://github.com/kata-containers/kata-containers/issues/2157 to ensure we keep track of that.

I wonder if a well documented notation of "-annotation" would be enough to filter out the annotations we don't like to have enabled.  Moreover, the "-" could have precedence and we don't have to add yet another new option.

File: src/runtime/Makefile (hunk 1 above)
Comment (fidencio) [reply src/runtime/Makefile:162]: Will do, I'll get back to this one Today and address all the comments.
Thanks for the reviews!

**Comment by amshinde** [comment]
@jodh-intel @fidencio I agree its safer to have a list of allowed annotations, but currently we support quite a lot of annotations with only a handful of them that need to be disabled. My thinking in maintaining a list of disabled annotations was to keep the config file "relatively" clutter free.

**Comment by fidencio** [comment]
@amshinde, I get your point and I'll address that as part of https://github.com/kata-containers/kata-containers/issues/2157.  

There we should discuss what's the best approach, whether we could have two lists (allow & deny), or even a single allow list with a "-foo" notationt to disable "foo" (I can see it being used as: "-foo", ".*" ... as in, allow everything but foo).

Would be okay for you if we take that as part of #2157?

**Comment by fidencio** [comment]
Howdy!

I'd like to ask you to rebase your PR atop of current `main`, as https://github.com/kata-containers/kata-containers/pull/2197 has been merged and it has a fix that unblocks the CI.

Please, do the rebase and feel free `@fidencio` in case you do not have the powers to run `/test`.

Sorry for the inconvenience and happy hacking!

**Comment by fidencio** [comment]
Hey, we've had yet another breakage in our CI and now it's fixed as part of https://github.com/kata-containers/kata-containers/pull/2203 ... it means that you'll need to rebase your PR atop of current main.

Sorry for the inconvenience.

**Comment by fidencio** [comment]
As per the conversation in the Architecture Committee meeting Yesterday, I'm closing this PR and we'll be working on a better for users adding changes to the default configuration.

Conversation was good and we'll be able to come up with better solutions in the future.

**Pull request closed without merging**